### PR TITLE
[FEATURE] Ajoute un endpoint pour récupérer les résultats de quêtes dans le cadre d'une campagne (PIX-14837).

### PR DESCRIPTION
**Pull request by frinyvonnick**
## :unicorn: Problème
On a besoin d'afficher en fin de parcours un nouveau bloc pour informer l'utilisateur le statut de son attestation (ontenu, non obtenu, en cours de calcul). On a besoin que l'API nous fournisse un endpoint pour savoir si on doit afficher ce bloc ou non.

## :robot: Proposition
On propose d'ajouter un endpoint `/api/campaign-participations/{campaignParticipationId}/quest-results` qui prends la campaignParticipationId en params et qui utilise le userId de l'utilisateur connecté

## :rainbow: Remarques
On a ajouté une API interne `reward-api` 
On a crée un prehandler qui vérifie que la campaignParticipation passée dans l'url appartient bien à l'utilisateur connecté

## :100: Pour tester
🤔 
Tests Ok ? ✅

### CODE REVIEW

File: api/tests/profile/unit/domain/usecases/get-reward-by-id-and-type_test.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { getRewardByIdAndType } from '../../../../../src/profile/domain/usecases/get-reward-by-id-and-type.js';
+import { expect, sinon } from '../../../../test-helper.js';
+
+describe('Profile | Unit | UseCase | get-reward-by-id-and-type', function () {
```
Comment (Libouk) [inline api/tests/profile/unit/domain/usecases/get-reward-by-id-and-type_test.js:4]: given when then

File: api/tests/profile/unit/application/api/reward-api_test.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { getByIdAndType } from '../../../../../src/profile/application/api/reward-api.js';
+import { usecases } from '../../../../../src/profile/domain/usecases/index.js';
+import { expect, sinon } from '../../../../test-helper.js';
+
+describe('Profile | Unit | Application | Api | reward', function () {
```
Comment (Libouk) [inline api/tests/profile/unit/application/api/reward-api_test.js:5]: given when then

File: api/tests/profile/integration/infrastructure/repositories/reward-repository_test.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      expect(result.id).to.equal(reward.id);
+      expect(result.key).to.equal(reward.key);
+    });
+
+    it('should throw RewardTypeDoesNotExistError if rewardType does not exist', async function () {
+      // given&when
+      const error = await catchErr(getByIdAndType)({ rewardId: 1, rewardType: 'NOT_EXISTING_REWARD_TYPE' });
+
+      // then
+
```
Comment (Libouk) [inline api/tests/profile/integration/infrastructure/repositories/reward-repository_test.js]: les sauts de ligne

File: api/src/quest/infrastructure/repositories/reward-repository.js [hunk 4]
```diff
@@ -1,7 +1,46 @@
...
+
+export const getByQuestAndUserId = async ({
+  userId,
+  quest,
+  rewardApi,
+  profileRewardApi,
+  profileRewardTemporaryStorage,
+}) => {
+  const reward = await rewardApi.getByIdAndType({ rewardId: quest.rewardId, rewardType: quest.rewardType });
+  const profileRewards = await profileRewardApi.getByUserId(userId);
```
Comment (Libouk) [inline api/src/quest/infrastructure/repositories/reward-repository.js:19]: vérifier dans le cas où il n'y a pas de profile rewards le comportement

File: api/tests/quest/unit/infrastructure/repositories/reward-repository_test.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      profileRewardTemporaryStorageStub = {
+        get: sinon.stub(),
+      };
+
+      rewardApiStub.getByIdAndType
+        .withArgs({ rewardId: quest.rewardId, rewardType: quest.rewardType })
+        .resolves(reward);
+    });
+
+    it('should return QuestResult with obtained true when reward is obtained', async function () {
```
Comment (Libouk) [inline api/tests/quest/unit/infrastructure/repositories/reward-repository_test.js:37]: given when then

File: api/tests/quest/unit/domain/models/Eligibility_test.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import { Eligibility } from '../../../../../src/quest/domain/models/Eligibility.js';
+import { expect } from '../../../../test-helper.js';
+
+describe('Quest | Unit | Domain | Models | Eligibility ', function () {
+  describe('#campaignParticipations', function () {
+    it('Should return an objet with targetProfileIds property', function () {
```
Comment (Libouk) [inline api/tests/quest/unit/domain/models/Eligibility_test.js]: object

File: api/src/quest/domain/usecases/get-quest-results-for-campaign-participation.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  userId,
+  campaignParticipationId,
+  questRepository,
+  eligibilityRepository,
+  rewardRepository,
+}) => {
+  const quests = await questRepository.findAll();
+
+  if (quests.length === 0) {
+    return false;
```
Comment (Libouk) [inline api/src/quest/domain/usecases/get-quest-results-for-campaign-participation.js]: return [] plutôt que false

File: api/tests/quest/integration/domain/usecases/get-quest-results-for-campaign-participation_test.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { COMPARISON } from '../../../../../src/quest/domain/models/Quest.js';
+import { QuestResult } from '../../../../../src/quest/domain/models/QuestResult.js';
+import { usecases } from '../../../../../src/quest/domain/usecases/index.js';
+import { databaseBuilder, expect } from '../../../../test-helper.js';
+
+describe('Quest | Integration | Domain | Usecases | getQuestResultsForCampaignParticipation', function () {
```
Comment (Libouk) [inline api/tests/quest/integration/domain/usecases/get-quest-results-for-campaign-participation_test.js:6]: Manquent les tests des edge cases

File: api/src/prescription/campaign/domain/errors.js [hunk 9]
```diff
@@ -52,9 +52,16 @@ class DeletedCampaignError extends DomainError {
   }
 }
 
+class CampaignParticipationDoesNotBelongToUser extends DomainError {
+  constructor(message = "La participation n'est pas liée a l'utilisateur") {
```
Comment (Libouk) [inline api/src/prescription/campaign/domain/errors.js]: à

File: api/src/shared/application/usecases/checkCampaignParticipationBelongsToUser.js
Comment (Libouk) [inline api/src/shared/application/usecases/checkCampaignParticipationBelongsToUser.js:1]: A mettre dans un contexte prescription

File: api/tests/quest/acceptance/application/quest-route_test.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+      const { id: campaignParticipationId } = databaseBuilder.factory.buildCampaignParticipation({
+        organizationLearnerId,
+        userId,
+      });
+      const rewardId = databaseBuilder.factory.buildAttestation().id;
+      databaseBuilder.factory.buildQuest({
+        rewardType: 'attestations',
+        rewardId,
+        eligibilityRequirements: [
+          {
```
Comment (Libouk) [inline api/tests/quest/acceptance/application/quest-route_test.js]: Remplacer par un tableau vide ?

File: api/tests/quest/unit/application/index_test.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import * as moduleUnderTest from '../../../../src/quest/application/index.js';
+import { questController } from '../../../../src/quest/application/quest-controller.js';
+import { securityPreHandlers } from '../../../../src/shared/application/security-pre-handlers.js';
+import { expect, HttpTestServer, sinon } from '../../../test-helper.js';
+
+describe('Quest | Unit | Router | quest-router', function () {
+  describe('GET /api/campaign-participations/{campaignParticipationId}/quest-results', function () {
+    it('should return 204 if campaign participation belongs to user', async function () {
```
Comment (Libouk) [inline api/tests/quest/unit/application/index_test.js]: it should call participation belongs to user prehandler

File: api/src/quest/infrastructure/repositories/reward-repository.js (hunk 4 above)
Comment (Alexandre-Monney) [reply api/src/quest/infrastructure/repositories/reward-repository.js:19]: l'api renvoie bien un tableau vide, donc pas de soucis de ce côté la 👍

**Review by Alexandre-Monney — APPROVED** [review]
👍 🐈‍⬛
